compiler/src/iree/compiler/Codegen/LLVMGPU/AMDGPUChainedMatmulPass.cpp

```diff
@@ -13,70 +13,121 @@ namespace mlir::iree_compiler {
+  /// fused into the contract. `isOperandSwapInvariant` tells us when we can
+  /// simply swap the operands without transposing them.
   void swapOperandsAndTranspose(RewriterBase &rewriter,
                                 vector::ContractionOp contractOp) const {
     Value lhs = contractOp.getLhs();
     Value rhs = contractOp.getRhs();
     Value acc = contractOp.getAcc();
     rewriter.setInsertionPoint(contractOp);
-    Value transposed = rewriter.create<vector::TransposeOp>(
-        contractOp.getLoc(), acc, SmallVector<int64_t>{1, 0});
-    AffineExpr m, n, k;
-    bindDims(rewriter.getContext(), m, n, k);
-    using MapList = ArrayRef<ArrayRef<AffineExpr>>;
-    auto infer = [&](MapList m) {
-      return AffineMap::inferFromExprList(m, contractOp.getContext());
-    };
-    SmallVector<AffineMap> newIndexingMaps = infer({{n, k}, {m, k}, {n, m}});
+    acc = rewriter.create<vector::TransposeOp>(contractOp.getLoc(), acc,
+                                               SmallVector<int64_t>{1, 0});
+
+    if (!isOperandSwapInvariant(contractOp)) {
+      lhs = rewriter.create<vector::TransposeOp>(contractOp.getLoc(), lhs,
+                                                 SmallVector<int64_t>{1, 0});
+      rhs = rewriter.create<vector::TransposeOp>(contractOp.getLoc(), rhs,
+                                                 SmallVector<int64_t>{1, 0});
+    }
+
     vector::ContractionOp swappedOp = rewriter.create<vector::ContractionOp>(
-        contractOp.getLoc(), rhs, lhs, transposed,
-        rewriter.getAffineMapArrayAttr(newIndexingMaps),
+        contractOp.getLoc(), rhs, lhs, acc, contractOp.getIndexingMaps(),
         contractOp.getIteratorTypesAttr());
-    Value newResult = swappedOp.getResult();
-    transposed = rewriter.create<vector::TransposeOp>(
-        contractOp.getLoc(), newResult, SmallVector<int64_t>{1, 0});
-    rewriter.replaceAllUsesWith(contractOp.getResult(), transposed);
+    rewriter.replaceOpWithNewOp<vector::TransposeOp>(
+        contractOp, swappedOp.getResult(), SmallVector<int64_t>{1, 0});
   }
 
-  /// The only compatible indexing map corresponds to
-  /// the matmul_transpose_b, and is
-  /// (m, n, k) -> (m, k)
-  /// (m, n, k) -> (n, k)
-  /// (m, n, k) -> (m, n)
-  bool isCompatibleIndexingMap(vector::ContractionOp contractOp,
-                               MLIRContext *ctx) {
+  /// For a matmul_transpose_b, this transformation boils down to an operand
+  /// swap and result transpose:
+  ///
+  /// def matmul_transpose_b(A, B):
+  ///   B.T = transpose(B)
+  ///   C = A @ B.T
+  ///   return C
+  ///
+  /// def matmul_transpose_b_swapped(A, B):
+  ///   A.T = transpose(A)
+  ///   C.T = B @ A.T
+  ///   C   = transpose(C.T)
+  ///   return C
+  ///
+  /// matmul_transpose_b(B, A) = matmul_transpose_b_swapped(B, A).T
+  ///
+  /// TODO: This check applies more generally when one of the operands in the
+  /// function is transposed compared to what "@" expects.
```

Comment:
Could you add a comment for the regular matmul case
```
C = A @ B
```
In this case to get C.T we need to swap and transpose
```
C.T = B.T @ A.T
```

Reply:
Added